packages/aws_common/lib/src/logging/aws_logger.dart

```diff
@@ -72,10 +92,56 @@ class AWSLogger implements Closeable {
     return AWSLogger('$namespace.$name');
   }
 
+  /// Returns a plugin of type [Plugin] registered to this
+  /// logger hierarchy or `null`.
+  Plugin? getPlugin<Plugin extends AWSLoggerPlugin>() {
+    final registeredPlugin = _subscriptions.keys
+            .firstWhereOrNull((element) => element.runtimeType == Plugin)
+        as Plugin?;
+    return registeredPlugin ?? _parent?.getPlugin<Plugin>();
+  }
+
   /// Registers an [AWSLoggerPlugin] to handle logs emitted by this logger
   /// instance.
-  void registerPlugin(AWSLoggerPlugin plugin) {
-    unregisterPlugin(plugin);
+  ///
+  /// Throws [StateError] if a plugin with same type is registered to this
+  /// logger hierarchy.
+  void registerPlugin<T extends AWSLoggerPlugin>(
+    T plugin,
+  ) {
+    if (_subscriptions.keys.any((element) => element.runtimeType == T)) {
+      throw StateError(_stateErorrMsgPluginIsRegistered(this));
+    }
+    final queue = Queue<AWSLogger>();
+
+    if (_parent != null) {
+      queue.add(_parent!);
+    }
+    while (queue.isNotEmpty) {
+      final logger = queue.removeFirst();
+      if (logger._subscriptions.keys
+          .any((element) => element.runtimeType == T)) {
+        throw StateError(_stateErorrMsgPluginIsRegistered(logger));
+      }
+      if (logger._parent != null) {
+        queue.add(logger._parent!);
+      }
+    }
+
+    if (_children.isNotEmpty) {
+      queue.addAll(_children);
+    }
+    while (queue.isNotEmpty) {
+      final logger = queue.removeFirst();
+      if (logger._subscriptions.keys
+          .any((element) => element.runtimeType == T)) {
+        throw StateError(_stateErorrMsgPluginIsRegistered(logger));
+      }
+      if (logger._children.isNotEmpty) {
+        queue.addAll(logger._children);
+      }
+    }
```

Comment:
```suggestion
    bool hasPlugin(AWSLogger logger) =>
        logger.getPlugin<T>() != null || logger._children.any(hasPlugin);
    if (hasPlugin(this)) {
      throw StateError(_stateErorrMsgPluginIsRegistered(this));
    }
```

Reply:
the queue implementation has a fixed o(n) n = logger hierarchy depth. the recursive version is less effiecient (I think should be like o(n2)) think of a scenario that a plugin is already registered to a node in the logger hierarchy and then calling registere to the root logger.

Reply:
You're discounting the cost of creating/manipulating a Queue (which is backed by an array) which is O(n) in each usage here.

**My version**
Time complexity: `logger.getPlugin` is O(depth) for each child + children iteration is O(n) = O(n + dn) = O((d + 1) * n), which would simplify to O(n) for shallow trees.
Space complexity: O(1) (no data structures are initialized)

**Your version**
Time complexity: Each logger in the tree is visited, thus O(n + d) + time to manipulate queues (creation, adding, removing, and garbage collection) O(n) = O(n + d) but with a large factor on n.
Space complexity: O(n) for the queues

In most use cases, I would expect my version to perform better. And even if it were marginally slower, this method is not expected to be called often and the complexity of your code is not warranted.

Reply:
thanks for hashing out the details. this is the summary of comparing the two:

**Time complexity**
n= number of loggers in the hierarchy
Queue implementation: o(n)
Recursive implementation: o(nlogn)

**Space complexity**
Queue implementation can be improved by using `ListQueue<AWSLogger>(activeLoggers.length);` instead of `Queue<AWSLogger>()`. 
Recusrsive version uses Stack space.

I have a slight preference for using queue. happy to change the code if you think we should do recursive.

Reply:
The recursive code is much simpler to read and understand and avoids duplication of logic. If you want to clean up and optimize the iterative approach, then please do so, but it is currently slower in most cases and reads worse.

The recursive can trivially be made O(n), thus being quicker in all cases than the iterative as well.

```dart
bool hasPlugin(AWSLogger logger) =>
    logger._subscriptions.keys.any((element) => element.runtimeType == T) ||
    logger._children.any(hasPlugin);
if (getPlugin<T>() != null || _children.any(hasPlugin)) {
  throw StateError(_stateErorrMsgPluginIsRegistered(this));
}
```